edx_event_bus_kafka/internal/producer.py

```diff
@@ -210,6 +219,46 @@ def prepare_for_shutdown(self):
         self.producer.flush(-1)
 
 
+def poll_indefinitely(api_weakref: EventProducerKafka):
+    """
+    Poll the producer indefinitely to ensure delivery/stats/etc. callbacks are triggered.
+
+    The thread stops automatically once the producer is garbage-collected.
+
+    This ensures that callbacks are triggered in a timely fashion, rather than waiting
+    for the poll() call that we make before or after each produce() call. This may be
+    important if events are produced infrequently, and it allows the last event the
+    server emits before shutdown to have its callback run (if it happens soon enough.)
+    """
+    # The reason we hold a weakref to the whole EventProducerKafka and
+    # not directly to the Producer itself is that you just can't make
+    # a weakref to the latter (perhaps because it's a C object.)
+
+    # .. setting_name: EVENT_BUS_KAFKA_POLL_INTERVAL_SEC
+    # .. setting_default: 1.0
+    # .. setting_description: How frequently to poll the event-bus-kafka producer. This should
+    #   be small enough that there's not too much latency in triggering delivery callbacks once
+    #   a message has been acknowledged, but there's no point in setting it any lower than the
+    #   expected round-trip-time of message delivery and acknowledgement. (100 ms – 5 s is
+    #   probably a reasonable range.)
+    poll_interval_seconds = getattr(settings, 'EVENT_BUS_KAFKA_POLL_INTERVAL_SEC', 1.0)
+    while True:
+        time.sleep(poll_interval_seconds)
+
+        # Temporarily hold a strong ref to the producer API singleton
+        api_object = api_weakref()
+        if api_object is None:
+            return
+
+        try:
+            api_object.producer.poll(0)
+        except BaseException:
+            pass
```

Comment:
Should we log the exception here?

Reply:
I'm a little concerned about logging so frequently if something goes wrong. But maybe it's not _that_ frequent?

Reply:
Hm. Maybe a debug log so we can see it if we make an effort but it won't log by default on most of our production servers?